src/Npgsql/NpgsqlTypes/NpgsqlCube.cs

```diff
@@ -0,0 +1,140 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+#pragma warning disable 1591
+
+// ReSharper disable once CheckNamespace
+namespace NpgsqlTypes
+{
+    public readonly struct NpgsqlCube : IEquatable<NpgsqlCube>
+    {
+        public const int MaxDimensions = 100;
+        
+        internal static readonly Regex Regex =
+            new(@"^(\[(?=\(.+\)\,\())?((\()?(-?\d+\.?\d*?(?:,-?\d+(?:\.?\d*))*)(?(3)\)|)(,\((-?\d+\.?\d*(?:,-?\d+\.?\d*)*)\))?)(?(1)\]|)$");
+        
+        /// <summary>
+        /// The lower left coordinates of the cube.
+        /// </summary>
+        public List<double> LowerLeft { get; }
+        
+        /// <summary>
+        /// The upper right coordinates of the cube.
+        /// </summary>
+        public List<double> UpperRight { get; }
+
+        /// <summary>
+        /// Constructs a point from coordinates.
+        /// </summary>
+        /// <param name="coords">The coordinates of the cube.</param>
+        public NpgsqlCube(IEnumerable<double> coords) : this(coords.ToArray()) { }
+        
+        /// <summary>
+        /// Constructs a point from coordinates.
+        /// </summary>
+        /// <param name="coords">The coordinates of the cube.</param>
+        public NpgsqlCube(params double[] coords) : this(coords, coords) { }
+
+        /// <summary>
+        /// Constructs a cube from coordinates.
+        /// </summary>
+        /// <param name="lowerLeft">The lower-left coordinates.</param>
+        /// <param name="upperRight">The upper-right coordinates.</param>
+        /// <exception cref="ArgumentException">
+        /// Thrown if the number of dimensions in the lower-left and upper-right coordinates are different.
+        /// </exception>
+        /// <exception cref="ArgumentOutOfRangeException">Thrown if the cube has too many dimensions.</exception>
+        public NpgsqlCube(IEnumerable<double> lowerLeft, IEnumerable<double> upperRight)
+        {
+            LowerLeft = lowerLeft.ToList();
+            UpperRight = upperRight.ToList();
+
+            if (UpperRight.Count != LowerLeft.Count)
+                throw new ArgumentException("Not a valid cube: Different point dimensions in " + ToString(), nameof(upperRight));
+            if (Dimensions() > MaxDimensions)
+                throw new ArgumentOutOfRangeException(nameof(lowerLeft), "Not a valid cube: Too many dimensions " + Dimensions());
+        }
+
+        /// <summary>
+        /// Returns the number of dimensions of the cube.
+        /// </summary>
+        /// <returns></returns>
+        public int Dimensions() => Math.Max(LowerLeft.Count, UpperRight.Count);
```

Comment:
@roji Maybe not, I'm not sure what the best approach is, but I think it might just be easier to fill in missing coordinates with zeros, so therefore the dimensions is just the length of the largest list. I think this makes the most sense since users can manipulate the lists however they want with the exposed mutable lists.

Reply:
That's certainly possible... I generally lean on the side of requiring users to provide correct data and not correcting it for them (e.g. by padding zeros) - that can help catch bugs in user code.

Reply:
If we do take that approach, where would you suggest throwing the exceptions? We can't throw when adding to a list because that wouldn't be possible, but it is possible to throw an exception when calling `IsPoint()` or `Dimensions()`. We could also only throw exceptions in the constructor and when attempting to write to the database.

Reply:
> We could also only throw exceptions in the constructor and when attempting to write to the database.

That seems pretty good. Though throwing from IsPoint and Dimensions also makes sense to me, since they can't return sensible results in that invalid state.

Reply:
They _can_ return a sensible result, but maybe they _shouldn't_ as you said. If we converted `IsPoint` and `Dimensions` into properties, then they would always be safe to call even if the data might not be perfect. However, if we keep them as methods, then it would be totally fine to throw an exception even if that inconveniences the user. I could really go either way as long as it produces a good experience.

Reply:
These do look like they'd be better off as properties, and I don't think there's any particular problem with throwing exceptions from properties.

Reply:
According to Microsoft, we shouldn't throw exceptions on property getters https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property. However, maybe we can make an exception (pun intended).

Reply:
Yeah, that's one of those guidelines that I'm a bit mitigated about :wink: If it's in the user's best interest that the property throw for invalid state, then I think we should do so, as it allows them to detect errors early. That doesn't seem like a good reason to turn the property into a method, which would penalize everyone else using it correctly...